Link your codeforces profile with cli

ashutoshsuthar2020 opened the pull request:
Added a profile command which allows you to make profile card of the given user
For example 

ecp profile tourist
*-*-*-*-*-*-**-*-*-*-*-*-*-*-*-*-*-*-*-*-*-
User Name : GennadyKorotkevich 
Rating : 3834 
Contribution : 134 
Rank : legendary grandmaster 
Max Rating : 3979
*-*-*-*-*-*-**-*-*-*-*-*-*-*-*-*-*-*-*-*-*-

ashutoshsuthar2020 committed:
Link your codeforces profile with cli

akstron commented:
Kindly provide description for each PR you raise.

akstron commented:
By linking I meant that we can store current user info somewhere. Your are able to display some random user as of now. Can we store that so that we can use this info somewhere in other commands. Also, I think that this command should be included in config command group. Kindly put your arguments of creating a new command group of it.

ashutoshsuthar2020 committed:
added necessary chancges to link wiith cofeforces profile

ashutoshsuthar2020 commented:
Sir please check now i have created function to save and fetch data about user in config file

ashutoshsuthar2020 committed:
Merge branch 'ContriHUB:main' into main

akstron commented on src/Config/Config.py:
```diff
@@ -88,3 +89,35 @@ def remove_proxy(self):
+        config_file_path = self.config_file_path
+        with open(config_file_path, 'r+') as config_file:
+            config = json.load(config_file)
+            config["firstname"] = html_content["result"][0]["firstName"]
+            config["lastname"] = html_content["result"][0]["lastName"]
+            config["rating"] = str(html_content["result"][0]["rating"]),
+            config["contri"] = str(html_content["result"][0]["contribution"]),
+            config["rank"] = str(html_content["result"][0]["rank"]),
+            config["maxrating"] = str(html_content["result"][0]["maxRating"])
```
We can create a single object with name profile, and put above as key value pairs, and then write that object.

akstron commented on COMMANDS.md:
````diff
@@ -20,6 +20,24 @@
     This will change the runtime environment. 
     Currently value of `<language>` can only be 'python' or 'cpp' are supported
 
+*   ```
+    ecp config set_user <username>
+    ```
````
Shouldn't it be 'set-user' instead of 'set_user'?

ashutoshsuthar2020 committed:
made necessary changes to Config.py

ashutoshsuthar2020 committed:
Update config.json

ashutoshsuthar2020 committed:
Update COMMANDS.md

ashutoshsuthar2020 commented:
Have made the recommended changes

ashutoshsuthar2020 commented:
src/Cli/config.py have called request from self.get_proxy() ,and it works properly.

akstron commented on src/Config/Config.py:
```diff
@@ -88,3 +89,27 @@ def remove_proxy(self):
             config_file.write(json.dumps(config))
+
+    def set_user(self, user):
+        response = requests.get(url='http://codeforces.com/api/user.info?handles='+user,proxies=self.get_proxy())
```
We can store the url in a variable first and then use it to improve readability

akstron replied:
Not done yet

ashutoshsuthar2020 committed:
Update Config.py

ashutoshsuthar2020 commented:
done

akstron commented on src/Cli/config.py:
```diff
@@ -66,8 +66,27 @@ def remove_proxy():
+        click.echo(e)
+@click.command()
+def get_user():
+    try:
+        config = Config()
+        config.get_user()
```
It's not printing anything

akstron commented on src/Config/Config.py:
```diff
@@ -88,3 +89,28 @@ def remove_proxy(self):
+        address='http://codeforces.com/api/user.info?handles='+user
+        response = requests.get(url=address,proxies=self.get_proxy())
+        if(response.status_code!=200):
+            raise UsernameError('User not Found')
```
No implementation of UsernameError class. Also I think "UsernameNotFound" would be a better name

ashutoshsuthar2020 committed:
Update Config.py

ashutoshsuthar2020 committed:
Update config.py

akstron commented on src/Cli/config.py:
```diff
@@ -79,7 +79,7 @@ def set_user(user):
     try:
         config = Config()
-        config.get_user()
+        print(config.get_user())
```
Use click.echo with styling.

akstron commented on src/Config/Config.py:
```diff
@@ -94,7 +94,7 @@ def set_user(self, user):
         response = requests.get(url=address,proxies=self.get_proxy())
         if(response.status_code!=200):
-            raise UsernameError('User not Found')
+            raise UsernameError('UsernameNotFound')
```
I was saying that there is no such class as "UsernameError". You have to create it on your own. Also I think the name of class should be "UsernameNotFound" with error string as "User not found".

ashutoshsuthar2020 committed:
added correct implemenatation of exception

akstron commented on src/Config/Config.py:
```diff
@@ -7,6 +7,10 @@
 '''
     Class to manage config related commands
 '''
+class UsernameNotFound(Exception):
+    def __init__(self, msg):
+        super().__init__(self, msg)
```
In the previous review as well I said shouldn't it have a default msg = "User not found"?
Also, kindly make a separate directory for exceptions related to profile and write class over there. Look into "Runner" directory, you can find an example there.

ashutoshsuthar2020 committed:
organised exception in config.py

akstron commented:
When using "get-user" we  can raise an exception when user is not set earlier.

ashutoshsuthar2020 committed:
UserNotSet exception class created

ashutoshsuthar2020 commented:
When using "get-user" we can now raise an exception when user is not set earlier.

ashutoshsuthar2020 committed:
api url varible formed

akstron commented on src/Config/Config.py:
```diff
@@ -88,3 +91,29 @@ def remove_proxy(self):
+        with open(config_file_path, 'r') as config_file:
+            config = json.load(config_file)
+        if(len(config["user"]["firstname"])==0):
+            raise UserNotSet
```
Are you not testing your commands after writing code?? Why raising a class?? 

Also, put proper spacing between functions and wherever you find it necessary.

ashutoshsuthar2020 committed:
modified UserNotSet.py and UserNotFound.py
and
fixed error
HTTPSConnectionPool(host='codeforces.com', port=443): Max retries exceeded with url: /api/user.info?handles=touris
(Caused by ProxyError('Your proxy appears to only use HTTP and not HTTPS, try changing your proxy URL to be HTTP.
See: https://urllib3.readthedocs.io/en/1.26.x/advanced-usage.html#https-proxy-error-http-proxy',
SSLError(SSLError(1, '[SSL: WRONG_VERSION_NUMBER] wrong version number (_ssl.c:1129)'))))

ashutoshsuthar2020 commented:
Sir check the changes made now.

akstron commented:
Receiving a proxy authentication error. Are you facing the same problem?

ashutoshsuthar2020 committed:
fixed the error on no proxy
reference : https://stackoverflow.com/questions/28521535/requests-how-to-disable-bypass-proxy

ashutoshsuthar2020 commented:
It was due to an update in the requests library, I have made the necessary changes.

akstron commented on src/Config/config.json:
```diff
@@ -1 +1 @@
-{"test_in_progress": "False", "language": "python", "template": "D:\\Alok country\\Apun coder banega\\Projects\\CP-CLI\\test\\code.cpp", "proxy": ""}
\ No newline at end of file
+{"user": {"firstname": "", "lastname": "", "rating": [""], "contri": [""], "rank": [""], "maxrating": ""}, "test_in_progress": "False", "language": "python", "template": "D:\\Alok country\\Apun coder banega\\Projects\\CP-CLI\\test\\code.cpp", "proxy": ""}
```
Why storing rating, contri, rank and maxrating in an array?

ashutoshsuthar2020 replied:
it will get easier if we call get-user method anywhere else in the code

akstron replied:
I don't get it. How come it will become easier?

ashutoshsuthar2020 replied:
I just tried to encapsulate user in one key value pair.
I thought it be easier to call property of user from this key value pair, instead of calling individual properties of user

akstron replied:
It's a good approach. Probably you are not getting my question. What I an asking is why is your rating property in user object an array? As you can see it's written that "rating": [""]. Why this instead of "rating" : "".

ashutoshsuthar2020 replied:
I had the same doubt actually, when I call it, it comes in this form only.

ashutoshsuthar2020 committed:
Update UserNotSet.py

ashutoshsuthar2020 committed:
Update config.py

ashutoshsuthar2020 committed:
Update config.py

ashutoshsuthar2020 reviewed (commented):
Applied recommended changes

akstron commented on src/Config/Config.py:
```diff
@@ -88,3 +92,30 @@ def remove_proxy(self):
+        config_file_path = self.config_file_path
+        with open(config_file_path, 'r+') as config_file:
+            config = json.load(config_file)
+            config["user"]["firstname"] = html_content["result"][0]["firstName"]
+            config["user"]["lastname"] = html_content["result"][0]["lastName"]
+            config["user"]["rating"] = str(html_content["result"][0]["rating"]),
+            config["user"]["contri"] = str(html_content["result"][0]["contribution"]),
+            config["user"]["rank"] = str(html_content["result"][0]["rank"]),
+            config["user"]["maxrating"] = str(html_content["result"][0]["maxRating"])
```
There's an error over here which is causing creation of an list. Try to figure it out.

ashutoshsuthar2020 replied:
The commas :cry: 
fixed it

ashutoshsuthar2020 committed:
Update Config.py

ashutoshsuthar2020 committed:
Update Config.py

ashutoshsuthar2020 committed:
Update config.json

akstron commented on src/Cli/config.py:
```diff
@@ -66,8 +66,27 @@ def remove_proxy():
+def get_user():
+    try:
+        config = Config()
+        click.echo(config.get_user())
```
As asked in the last review as well, why are you directly printing the object. Can you consider formatting it?

ashutoshsuthar2020 committed:
Update Config.py

ashutoshsuthar2020 committed:
Update config.py

akstron commented on src/Config/Config.py:
```diff
@@ -118,4 +118,5 @@ def get_user(self):
             raise UserNotSet()
-        return config["user"]
+        for info in config["user"]:
+            print(info + " : " + str(config["user"][info]))
```
Earlier approach was correct. Receive user object from Config.py and print it in config.py
